workspaces/diff-engine/src/queries/shape.rs

```diff
@@ -169,6 +306,43 @@ impl<'a> ShapeQueries<'a> {
     }
   }
 
+  pub fn resolve_parameters_to_shapes(
+    &self,
+    shape_id: &ShapeId,
+  ) -> Vec<(ShapeParameterId, ShapeId)> {
+    let projection = &self.shape_projection;
+
+    let shape_node_index = projection
+      .get_shape_node_index(shape_id)
+      .expect("shape id to resolve parameter for must exist");
+
+    projection
+      .graph
+      .neighbors_directed(*shape_node_index, petgraph::Direction::Incoming)
+      .filter_map(|neighbor_node_index| {
+        let neighbor = projection.graph.node_weight(neighbor_node_index).unwrap();
+        match neighbor {
+          Node::ShapeParameter(parameter_node) => {
+            let mut outgoing_edges = projection
+              .graph
+              .edges_connecting(*shape_node_index, neighbor_node_index);
```

Comment:
I wonder if instead of going straight to `neighbours` and then use `edges_connecting` to get back to the edge, it wouldn't make more sense to walk the edges first and then follow the edge. Maybe another efficiency gain that's not really worth anything.

Reply:
I was just lazy copy+pasting the singular one above - will update!